File: plugins/plugin-agent-orchestrator/python/elizaos_plugin_agent_orchestrator/actions/task_management.py
"""
Task management actions for the Agent Orchestrator plugin.
"""

from __future__ import annotations

import re
from typing import Any

from ..service import AgentOrchestratorService


def _get_service(runtime: Any) -> AgentOrchestratorService:
    """Get the orchestrator service from runtime."""
    svc = runtime.get_service("CODE_TASK")
    if svc is None:
        raise RuntimeError("AgentOrchestratorService not available (CODE_TASK)")
    return svc


def _extract_query(text: str) -> str:
    """Extract search query from text."""
    result = text.lower()
    result = re.sub(
        r"\b(switch|select|go|change|search|find|pause|stop|halt|resume|restart|continue|start|run|begin|cancel|delete|remove|list|show|view)\b",
        "",
        result,
    )
    result = re.sub(r"\b(about|for|named|called|with|to|my|your|our|this|current)\b", "", result)
    result = re.sub(r"\b(task|tasks|the|a|an)\b", "", result)
    result = re.sub(r"\s+", " ", result)
    return result.strip()


# ============================================================================
# CREATE_TASK
# ============================================================================

create_task_action = {
    "name": "CREATE_TASK",
    "similes": ["START_TASK", "SPAWN_TASK", "NEW_TASK", "BEGIN_TASK"],
    "description": "Create an orchestrated background task to be executed by a selected agent provider.",
}


async def validate_create_task(runtime: Any, message: Any) -> bool:
    """Validate CREATE_TASK action."""
    text = (getattr(message.content, "text", "") or "").lower()
    has_explicit = "create task" in text or "new task" in text or "start a task" in text
    has_intent = any(
        word in text
        for word in ["implement", "build", "create", "develop", "refactor", "fix", "add"]
    )
    return has_explicit or has_intent


async def handle_create_task(
    runtime: Any,
    message: Any,
    state: Any | None = None,
    options: dict[str, Any] | None = None,
    callback: Any | None = None,
) -> dict[str, Any]:
    """Handle CREATE_TASK action."""
    svc = _get_service(runtime)
    raw = getattr(message.content, "text", "") or ""

    opts = options or {}
    name = (opts.get("title") or raw.split("\n")[0] or "New Task").strip()[:100] or "New Task"
    description = (opts.get("description") or raw).strip()[:4000] or name

    room_id = getattr(message, "room_id", None) or getattr(message, "roomId", None)
    task = await svc.create_task(name, description, room_id)

    step_lines = opts.get("steps")
    if isinstance(step_lines, list) and step_lines:
        for s in step_lines:
            step = str(s).strip()
            if step:
                await svc.add_step(task.id, step)
        await svc.append_output(
            task.id, "Plan:\n" + "\n".join(f"{i + 1}. {s}" for i, s in enumerate(step_lines))
        )

    msg = (
        f"Created task: {task.name}\n"
        f"Provider: {task.metadata.provider_label or task.metadata.provider_id}\n"
        f"Starting execution…"
    )

    if callback:
        await callback({"content": {"text": msg}})

    svc.start_task_execution(task.id)

    return {"success": True, "text": msg, "data": {"taskId": task.id}}


create_task_action["validate"] = validate_create_task
create_task_action["handler"] = handle_create_task


# ============================================================================
# LIST_TASKS
# ============================================================================

list_tasks_action = {
    "name": "LIST_TASKS",
    "similes": ["SHOW_TASKS", "GET_TASKS", "TASKS", "VIEW_TASKS"],
    "description": "List tasks managed by the orchestrator.",
}


async def validate_list_tasks(runtime: Any, message: Any) -> bool:
    """Validate LIST_TASKS action."""
    text = (getattr(message.content, "text", "") or "").lower()
    return "list task" in text or "show task" in text or text == "tasks" or "my task" in text


async def handle_list_tasks(
    runtime: Any,
    message: Any,
    state: Any | None = None,
    options: dict[str, Any] | None = None,
    callback: Any | None = None,
) -> dict[str, Any]:
    """Handle LIST_TASKS action."""
    svc = _get_service(runtime)
    tasks = await svc.get_recent_tasks(20)

    if not tasks:
        msg = "No tasks."
        if callback:
            await callback({"content": {"text": msg}})
        return {"success": True, "text": msg}

    lines = ["Tasks:"]
    current = svc.get_current_task_id()
    for t in tasks:
        marker = " (current)" if t.id == current else ""
        lines.append(f"- {t.name} — {t.metadata.status.value} {t.metadata.progress}%{marker}")

    msg = "\n".join(lines)
    if callback:
        await callback({"content": {"text": msg}})
    return {"success": True, "text": msg}


list_tasks_action["validate"] = validate_list_tasks
list_tasks_action["handler"] = handle_list_tasks


# ============================================================================
# SWITCH_TASK
# ============================================================================

switch_task_action = {
    "name": "SWITCH_TASK",
    "similes": ["SELECT_TASK", "SET_TASK", "CHANGE_TASK", "GO_TO_TASK"],
    "description": "Switch the current task context to a different task.",
}


async def validate_switch_task(runtime: Any, message: Any) -> bool:
    """Validate SWITCH_TASK action."""
    text = (getattr(message.content, "text", "") or "").lower()
    return (
        "switch to task" in text or "select task" in text or ("task" in text and "switch" in text)
    )


async def handle_switch_task(
    runtime: Any,
    message: Any,
    state: Any | None = None,
    options: dict[str, Any] | None = None,
    callback: Any | None = None,
) -> dict[str, Any]:
    """Handle SWITCH_TASK action."""
    svc = _get_service(runtime)
    query = _extract_query(getattr(message.content, "text", "") or "")

    if not query:
        msg = "Please specify which task to switch to (by name or id)."
        if callback:
            await callback({"content": {"text": msg}})
        return {"success": False, "text": msg}

    matches = await svc.search_tasks(query)
    chosen = matches[0] if matches else None

    if not chosen or not chosen.id:
        msg = f'No task found matching: "{query}"'
        if callback:
            await callback({"content": {"text": msg}})
        return {"success": False, "text": msg}

    svc.set_current_task(chosen.id)
    msg = f"Switched to task: {chosen.name}"
    if callback:
        await callback({"content": {"text": msg}})
    return {"success": True, "text": msg, "data": {"taskId": chosen.id}}


switch_task_action["validate"] = validate_switch_task
switch_task_action["handler"] = handle_switch_task


# ============================================================================
# SEARCH_TASKS
# ============================================================================

search_tasks_action = {
    "name": "SEARCH_TASKS",
    "similes": ["FIND_TASK", "LOOKUP_TASK"],
    "description": "Search tasks by query.",
}


async def validate_search_tasks(runtime: Any, message: Any) -> bool:
    """Validate SEARCH_TASKS action."""
    text = (getattr(message.content, "text", "") or "").lower()
    return "search task" in text or "find task" in text or "look for task" in text


async def handle_search_tasks(
    runtime: Any,
    message: Any,
    state: Any | None = None,
    options: dict[str, Any] | None = None,
    callback: Any | None = None,
) -> dict[str, Any]:
    """Handle SEARCH_TASKS action."""
    svc = _get_service(runtime)
    opts = options or {}
    query = (
        opts.get("query") or _extract_query(getattr(message.content, "text", "") or "")
    ).strip()

    if not query:
        msg = "What would you like to search for?"
        if callback:
            await callback({"content": {"text": msg}})
        return {"success": False, "text": msg}

    matches = await svc.search_tasks(query)

    if not matches:
        msg = f'No tasks found matching: "{query}"'
        if callback:
            await callback({"content": {"text": msg}})
        return {"success": True, "text": msg}

    lines = [f'Found {len(matches)} task(s) matching "{query}":']
    for t in matches[:10]:
        lines.append(f"- {t.name} — {t.metadata.status.value} {t.metadata.progress}%")

    msg = "\n".join(lines)
    if callback:
        await callback({"content": {"text": msg}})
    return {"success": True, "text": msg}


search_tasks_action["validate"] = validate_search_tasks
search_tasks_action["handler"] = handle_search_tasks


# ============================================================================
# PAUSE_TASK
# ============================================================================

pause_task_action = {
    "name": "PAUSE_TASK",
    "similes": ["STOP_TASK", "HALT_TASK"],
    "description": "Pause a running task.",
}


async def validate_pause_task(runtime: Any, message: Any) -> bool:
    """Validate PAUSE_TASK action."""
    text = (getattr(message.content, "text", "") or "").lower()
    return any(word in text for word in ["pause", "stop", "halt"]) and "task" in text


async def handle_pause_task(
    runtime: Any,
    message: Any,
    state: Any | None = None,
    options: dict[str, Any] | None = None,
    callback: Any | None = None,
) -> dict[str, Any]:
    """Handle PAUSE_TASK action."""
    svc = _get_service(runtime)
    query = _extract_query(getattr(message.content, "text", "") or "")
    task = (await svc.search_tasks(query))[0] if query else await svc.get_current_task()

    if not task or not task.id:
        msg = "No task to pause."
        if callback:
            await callback({"content": {"text": msg}})
        return {"success": False, "text": msg}

    await svc.pause_task(task.id)
    msg = f"Paused task: {task.name}"
    if callback:
        await callback({"content": {"text": msg}})
    return {"success": True, "text": msg}


pause_task_action["validate"] = validate_pause_task
pause_task_action["handler"] = handle_pause_task


# ============================================================================
# RESUME_TASK
# ============================================================================

resume_task_action = {
    "name": "RESUME_TASK",
    "similes": ["CONTINUE_TASK", "RESTART_TASK", "RUN_TASK"],
    "description": "Resume a paused task.",
}


async def validate_resume_task(runtime: Any, message: Any) -> bool:
    """Validate RESUME_TASK action."""
    text = (getattr(message.content, "text", "") or "").lower()
    return "task" in text and any(word in text for word in ["resume", "restart", "continue"])


async def handle_resume_task(
    runtime: Any,
    message: Any,
    state: Any | None = None,
    options: dict[str, Any] | None = None,
    callback: Any | None = None,
) -> dict[str, Any]:
    """Handle RESUME_TASK action."""
    svc = _get_service(runtime)
    query = _extract_query(getattr(message.content, "text", "") or "")
    task = (await svc.search_tasks(query))[0] if query else await svc.get_current_task()

    if not task or not task.id:
        msg = "No task to resume."
        if callback:
            await callback({"content": {"text": msg}})
        return {"success": False, "text": msg}

    await svc.resume_task(task.id)
    svc.start_task_execution(task.id)
    msg = f"Resumed task: {task.name}"
    if callback:
        await callback({"content": {"text": msg}})
    return {"success": True, "text": msg}


resume_task_action["validate"] = validate_resume_task
resume_task_action["handler"] = handle_resume_task


# ============================================================================
# CANCEL_TASK
# ============================================================================

cancel_task_action = {
    "name": "CANCEL_TASK",
    "similes": ["DELETE_TASK", "REMOVE_TASK", "ABORT_TASK"],
    "description": "Cancel a task.",
}


async def validate_cancel_task(runtime: Any, message: Any) -> bool:
    """Validate CANCEL_TASK action."""
    text = (getattr(message.content, "text", "") or "").lower()
    return any(word in text for word in ["cancel", "delete", "remove"]) and "task" in text


async def handle_cancel_task(
    runtime: Any,
    message: Any,
    state: Any | None = None,
    options: dict[str, Any] | None = None,
    callback: Any | None = None,
) -> dict[str, Any]:
    """Handle CANCEL_TASK action."""
    svc = _get_service(runtime)
    query = _extract_query(getattr(message.content, "text", "") or "")
    task = (await svc.search_tasks(query))[0] if query else await svc.get_current_task()

    if not task or not task.id:
        msg = "No task to cancel."
        if callback:
            await callback({"content": {"text": msg}})
        return {"success": False, "text": msg}

    await svc.cancel_task(task.id)
    msg = f"Cancelled task: {task.name}"
    if callback:
        await callback({"content": {"text": msg}})
    return {"success": True, "text": msg}


cancel_task_action["validate"] = validate_cancel_task
cancel_task_action["handler"] = handle_cancel_task

File: plugins/plugin-agent-orchestrator/typescript/src/actions/task-management.ts
import type {
  Action,
  ActionResult,
  HandlerCallback,
  HandlerOptions,
  IAgentRuntime,
  Memory,
  State,
} from "@elizaos/core";
import type { AgentOrchestratorService } from "../services/agent-orchestrator-service.js";

function getService(runtime: IAgentRuntime): AgentOrchestratorService {
  const svc = runtime.getService("CODE_TASK") as AgentOrchestratorService | null;
  if (!svc) {
    throw new Error("AgentOrchestratorService not available (CODE_TASK)");
  }
  return svc;
}

function extractQuery(text: string): string {
  return text
    .toLowerCase()
    .replace(
      /\b(switch|select|go|change|search|find|pause|stop|halt|resume|restart|continue|start|run|begin|cancel|delete|remove|list|show|view)\b/g,
      "",
    )
    .replace(/\b(about|for|named|called|with|to|my|your|our|this|current)\b/g, "")
    .replace(/\b(task|tasks|the|a|an)\b/g, "")
    .replace(/\s+/g, " ")
    .trim();
}

// ============================================================================
// CREATE_TASK
// ============================================================================

export const createTaskAction: Action = {
  name: "CREATE_TASK",
  similes: ["START_TASK", "SPAWN_TASK", "NEW_TASK", "BEGIN_TASK"],
  description:
    "Create an orchestrated background task to be executed by a selected agent provider.",
  validate: async (_runtime: IAgentRuntime, message: Memory): Promise<boolean> => {
    const text = message.content.text?.toLowerCase() ?? "";
    const hasExplicit =
      text.includes("create task") || text.includes("new task") || text.includes("start a task");
    const hasIntent =
      text.includes("implement") ||
      text.includes("build") ||
      text.includes("create") ||
      text.includes("develop") ||
      text.includes("refactor") ||
      text.includes("fix") ||
      text.includes("add");
    return hasExplicit || hasIntent;
  },
  handler: async (
    runtime: IAgentRuntime,
    message: Memory,
    _state?: State,
    options?: HandlerOptions,
    callback?: HandlerCallback,
  ): Promise<ActionResult> => {
    const svc = getService(runtime);
    const raw = message.content.text ?? "";

    const opts = options as { title?: string; description?: string; steps?: string[] } | undefined;
    const name =
      (opts?.title ?? raw.split("\n")[0] ?? "New Task").trim().slice(0, 100) || "New Task";
    const description = (opts?.description ?? raw).trim().slice(0, 4000) || name;

    const roomId = message.roomId;
    const task = await svc.createTask(name, description, roomId);

    const stepLines = Array.isArray(opts?.steps) ? opts?.steps : undefined;
    if (stepLines && stepLines.length > 0) {
      for (const s of stepLines) {
        const step = String(s).trim();
        if (!step) continue;
        await svc.addStep(task.id ?? "", step);
      }
      await svc.appendOutput(
        task.id ?? "",
        `Plan:\n${stepLines.map((s, i) => `${i + 1}. ${s}`).join("\n")}`,
      );
    }

    const msg = `Created task: ${task.name}\nProvider: ${task.metadata.providerLabel ?? task.metadata.providerId}\nStarting execution…`;
    await callback?.({ content: { text: msg } });

    svc.startTaskExecution(task.id ?? "").catch(() => {});

    return { success: true, text: msg, data: { taskId: task.id ?? "" } };
  },
};

// ============================================================================
// LIST_TASKS
// ============================================================================

export const listTasksAction: Action = {
  name: "LIST_TASKS",
  similes: ["SHOW_TASKS", "GET_TASKS", "TASKS", "VIEW_TASKS"],
  description: "List tasks managed by the orchestrator.",
  validate: async (_runtime: IAgentRuntime, message: Memory): Promise<boolean> => {
    const t = message.content.text?.toLowerCase() ?? "";
    return (
      t.includes("list task") || t.includes("show task") || t === "tasks" || t.includes("my task")
    );
  },
  handler: async (
    runtime: IAgentRuntime,
    _message: Memory,
    _state?: State,
    _options?: HandlerOptions,
    callback?: HandlerCallback,
  ): Promise<ActionResult> => {
    const svc = getService(runtime);
    const tasks = await svc.getRecentTasks(20);
    if (tasks.length === 0) {
      const msg = "No tasks.";
      await callback?.({ content: { text: msg } });
      return { success: true, text: msg };
    }
    const lines: string[] = ["Tasks:"];
    const current = svc.getCurrentTaskId();
    for (const t of tasks) {
      const marker = t.id === current ? " (current)" : "";
      lines.push(`- ${t.name} — ${t.metadata.status} ${t.metadata.progress}%${marker}`);
    }
    const msg = lines.join("\n");
    await callback?.({ content: { text: msg } });
    return { success: true, text: msg };
  },
};

// ============================================================================
// SWITCH_TASK
// ============================================================================

export const switchTaskAction: Action = {
  name: "SWITCH_TASK",
  similes: ["SELECT_TASK", "SET_TASK", "CHANGE_TASK", "GO_TO_TASK"],
  description: "Switch the current task context to a different task.",
  validate: async (_runtime: IAgentRuntime, message: Memory): Promise<boolean> => {
    const t = message.content.text?.toLowerCase() ?? "";
    return (
      t.includes("switch to task") ||
      t.includes("select task") ||
      (t.includes("task") && t.includes("switch"))
    );
  },
  handler: async (
    runtime: IAgentRuntime,
    message: Memory,
    _state?: State,
    _options?: HandlerOptions,
    callback?: HandlerCallback,
  ): Promise<ActionResult> => {
    const svc = getService(runtime);
    const query = extractQuery(message.content.text ?? "");
    if (!query) {
      const msg = "Please specify which task to switch to (by name or id).";
      await callback?.({ content: { text: msg } });
      return { success: false, text: msg };
    }
    const matches = await svc.searchTasks(query);
    const chosen = matches[0];
    if (!chosen?.id) {
      const msg = `No task found matching: "${query}"`;
      await callback?.({ content: { text: msg } });
      return { success: false, text: msg };
    }
    svc.setCurrentTask(chosen.id);
    const msg = `Switched to task: ${chosen.name}`;
    await callback?.({ content: { text: msg } });
    return { success: true, text: msg, data: { taskId: chosen.id } };
  },
};

// ============================================================================
// SEARCH_TASKS
// ============================================================================

export const searchTasksAction: Action = {
  name: "SEARCH_TASKS",
  similes: ["FIND_TASK", "LOOKUP_TASK"],
  description: "Search tasks by query.",
  validate: async (_runtime: IAgentRuntime, message: Memory): Promise<boolean> => {
    const t = message.content.text?.toLowerCase() ?? "";
    return t.includes("search task") || t.includes("find task") || t.includes("look for task");
  },
  handler: async (
    runtime: IAgentRuntime,
    message: Memory,
    _state?: State,
    options?: HandlerOptions,
    callback?: HandlerCallback,
  ): Promise<ActionResult> => {
    const svc = getService(runtime);
    const opt = options as { query?: string } | undefined;
    const query = (opt?.query ?? extractQuery(message.content.text ?? "")).trim();
    if (!query) {
      const msg = "What would you like to search for?";
      await callback?.({ content: { text: msg } });
      return { success: false, text: msg };
    }
    const matches = await svc.searchTasks(query);
    if (matches.length === 0) {
      const msg = `No tasks found matching: "${query}"`;
      await callback?.({ content: { text: msg } });
      return { success: true, text: msg };
    }
    const lines: string[] = [`Found ${matches.length} task(s) matching "${query}":`];
    for (const t of matches.slice(0, 10)) {
      lines.push(`- ${t.name} — ${t.metadata.status} ${t.metadata.progress}%`);
    }
    const msg = lines.join("\n");
    await callback?.({ content: { text: msg } });
    return { success: true, text: msg };
  },
};

// ============================================================================
// PAUSE / RESUME / CANCEL
// ============================================================================

export const pauseTaskAction: Action = {
  name: "PAUSE_TASK",
  similes: ["STOP_TASK", "HALT_TASK"],
  description: "Pause a running task.",
  validate: async (_runtime: IAgentRuntime, message: Memory): Promise<boolean> => {
    const t = message.content.text?.toLowerCase() ?? "";
    return (t.includes("pause") || t.includes("stop") || t.includes("halt")) && t.includes("task");
  },
  handler: async (
    runtime: IAgentRuntime,
    message: Memory,
    _state?: State,
    _options?: HandlerOptions,
    callback?: HandlerCallback,
  ): Promise<ActionResult> => {
    const svc = getService(runtime);
    const query = extractQuery(message.content.text ?? "");
    const task = query ? (await svc.searchTasks(query))[0] : await svc.getCurrentTask();
    if (!task?.id) {
      const msg = "No task to pause.";
      await callback?.({ content: { text: msg } });
      return { success: false, text: msg };
    }
    await svc.pauseTask(task.id);
    const msg = `Paused task: ${task.name}`;
    await callback?.({ content: { text: msg } });
    return { success: true, text: msg };
  },
};

export const resumeTaskAction: Action = {
  name: "RESUME_TASK",
  similes: ["CONTINUE_TASK", "RESTART_TASK", "RUN_TASK"],
  description: "Resume a paused task.",
  validate: async (_runtime: IAgentRuntime, message: Memory): Promise<boolean> => {
    const t = message.content.text?.toLowerCase() ?? "";
    return (
      t.includes("task") &&
      (t.includes("resume") || t.includes("restart") || t.includes("continue"))
    );
  },
  handler: async (
    runtime: IAgentRuntime,
    message: Memory,
    _state?: State,
    _options?: HandlerOptions,
    callback?: HandlerCallback,
  ): Promise<ActionResult> => {
    const svc = getService(runtime);
    const query = extractQuery(message.content.text ?? "");
    const task = query ? (await svc.searchTasks(query))[0] : await svc.getCurrentTask();
    if (!task?.id) {
      const msg = "No task to resume.";
      await callback?.({ content: { text: msg } });
      return { success: false, text: msg };
    }
    await svc.resumeTask(task.id);
    svc.startTaskExecution(task.id).catch(() => {});
    const msg = `Resumed task: ${task.name}`;
    await callback?.({ content: { text: msg } });
    return { success: true, text: msg };
  },
};

export const cancelTaskAction: Action = {
  name: "CANCEL_TASK",
  similes: ["DELETE_TASK", "REMOVE_TASK", "ABORT_TASK"],
  description: "Cancel a task.",
  validate: async (_runtime: IAgentRuntime, message: Memory): Promise<boolean> => {
    const t = message.content.text?.toLowerCase() ?? "";
    return (
      (t.includes("cancel") || t.includes("delete") || t.includes("remove")) && t.includes("task")
    );
  },
  handler: async (
    runtime: IAgentRuntime,
    message: Memory,
    _state?: State,
    _options?: HandlerOptions,
    callback?: HandlerCallback,
  ): Promise<ActionResult> => {
    const svc = getService(runtime);
    const query = extractQuery(message.content.text ?? "");
    const task = query ? (await svc.searchTasks(query))[0] : await svc.getCurrentTask();
    if (!task?.id) {
      const msg = "No task to cancel.";
      await callback?.({ content: { text: msg } });
      return { success: false, text: msg };
    }
    await svc.cancelTask(task.id);
    const msg = `Cancelled task: ${task.name}`;
    await callback?.({ content: { text: msg } });
    return { success: true, text: msg };
  },
};
